consensus/src/marshal/actor.rs

```diff
@@ -831,4 +935,90 @@ impl<
             Err(e) => panic!("failed to get block: {e}"),
         }
     }
+
+    /// Attempt to repair any identified gaps in the finalized blocks archive. The total
+    /// number of missing heights that can be repaired at once is bounded by `self.max_repair`,
+    /// though multiple gaps may be spanned.
+    async fn try_repair_gaps<K: PublicKey>(
+        &mut self,
+        buffer: &mut buffered::Mailbox<K, B>,
+        resolver: &mut impl Resolver<Key = Request<B>>,
+        application: &mut impl Reporter<Activity = Update<B>>,
+    ) {
+        for (gap_start, gap_end) in self.identify_gaps() {
+            // Attempt to repair the gap backwards from the end of the gap, using
+            // blocks from our local storage.
+            let Some(mut cursor) = self.get_finalized_block(gap_end).await else {
+                panic!("gapped block missing that should exist: {gap_end}");
+            };
+
+            // Iterate backwards, repairing blocks as we go.
+            while cursor.height() > gap_start {
+                let commitment = cursor.parent();
+                if let Some(block) = self.find_block(buffer, commitment).await {
+                    let finalization = self.cache.get_finalization_for(commitment).await;
+                    self.store_finalization(
+                        block.height(),
+                        commitment,
+                        block.clone(),
+                        finalization,
+                        application,
+                    )
+                    .await;
+                    debug!(height = block.height(), "repaired block");
+                    cursor = block;
+                } else {
+                    // Request the next missing block digest
+                    if !self
+                        .waiting_finalized
+                        .contains(&cursor.height().saturating_sub(1))
+                    {
+                        resolver.fetch(Request::<B>::Block(commitment)).await;
+                    }
+                    break;
+                }
+            }
+
+            // If we haven't fully repaired the gap, then also request any possible
+            // finalizations for the blocks in the remaining gap. This may help
+            // shrink the size of the gap if finalizations for the requests heights
+            // exist. If not, we rely on the recursive digest fetch above.
+            let gap_end = std::cmp::min(cursor.height(), gap_start);
```

Comment:
I think this prevents us from ever actually entering the loop below. I think you want `max` here?